File: app/techniques.py
"""Les techniques d'arts martiaux — les coups de rue et les cours du dojo.

Unités du moteur, comme `armes.py` : dégâts en points de vie, portée en pixels,
durées en images (60 par seconde). Une technique est une LIGNE DE TEMPS : une
suite d'étapes (une pose clé, sa durée, un décalage le long du regard `dx`, un
décalage vertical `dy`, une rotation `rot`, une hauteur `z`), dont UNE est
`actif` — c'est là que le coup porte, ou que la prise projette.

Les gestes (`static/js/techniques.js` choisit) :
- `tape` : FRAPPE en tapes ; `rang` est la place dans la chaîne (1, 2, 3…).
- `genou` : FRAPPE collé à la cible. `tenue` : FRAPPE tenue. `course` : FRAPPE
  en sprint. `roulade` : FRAPPE au sortir d'une roulade.
- `prise` : SAISIR relâché sans rien faire (repousser). `prise_frappe` : FRAPPE
  dans la prise. `prise_avant`, `prise_vers_soi`, `prise_cote` : le stick dans
  la prise, relatif à l'axe joueur → cible.
- `contre` : SAISIR pendant que l'ennemi arme son coup. `dos` : SAISIR tenu
  dans le dos de l'ennemi (`tenir` images).

`projete` : la distance du vol, en pixels (0 = on ne vole pas). `sans_sang` :
la victime ne saigne pas. `silencieuse` : ni cri, ni alerte autour.
"""

from __future__ import annotations

from typing import TypedDict

STYLES = ("rue", "boxe", "karate", "judo", "jiujitsu")
GESTES = ("tape", "genou", "tenue", "course", "roulade", "prise", "prise_frappe",
          "prise_avant", "prise_vers_soi", "prise_cote", "contre", "dos")
#: `debout` : la pose de marche (rien à dessiner) ; `frappe` : la pose de coup
#: qui existe déjà (`SPRITES.joueur.poses.frappe_*`). Les autres se dessinent
#: dans `sprites.js` sous `tech_<pose>_<bas|haut|cote>`.
POSES = ("debout", "frappe", "poing_arriere", "crochet", "genou", "pied_face", "pied_cote",
         "saisie", "hanche", "au_sol", "accroupi")
DESSINEES = POSES[2:]


class Etape(TypedDict):
    pose: str
    images: int
    dx: int
    dy: int
    rot: float
    z: int
    actif: bool


class Technique(TypedDict):
    slug: str
    nom: str
    style: str
    geste: str
    rang: int
    gratuite: bool
    prix: int
    degats: int
    portee: int
    arc: float
    renverse: bool
    assomme: bool
    projete: int
    tours: float
    silencieuse: bool
    sans_sang: bool
    tenir: int
    temps: list[Etape]


def _e(pose, images, *, dx=0, dy=0, rot=0.0, z=0, actif=False) -> Etape:
    return Etape(pose=pose, images=images, dx=dx, dy=dy, rot=rot, z=z, actif=actif)


def _t(slug, nom, style, geste, degats, portee, temps, *, rang=0, prix=0, arc=0.9,
       renverse=False, assomme=True, projete=0, tours=0.5, silencieuse=False,
       sans_sang=False, tenir=0) -> Technique:
    return Technique(slug=slug, nom=nom, style=style, geste=geste, rang=rang,
                     gratuite=style == "rue", prix=prix, degats=degats, portee=portee,
                     arc=arc, renverse=renverse, assomme=assomme, projete=projete,
                     tours=tours, silencieuse=silencieuse, sans_sang=sans_sang,
                     tenir=tenir, temps=temps)


CATALOGUE: list[Technique] = [
    # --- La rue : gratuite, pour tout le monde. Les poings de `armes.py` (8 de
    # dégâts, 12 px) restent la mesure : un direct EST le coup de poing d'avant.
    _t("direct_gauche", "Direct du gauche", "rue", "tape", 8, 12, rang=1, temps=[
        _e("debout", 3, dx=-1), _e("frappe", 4, dx=3, actif=True), _e("frappe", 3, dx=1), _e("debout", 2)]),
    _t("direct_droit", "Direct du droit", "rue", "tape", 8, 12, rang=2, temps=[
        _e("debout", 3, dx=-1), _e("poing_arriere", 4, dx=3, actif=True), _e("poing_arriere", 3, dx=1),
        _e("debout", 2)]),
    _t("crochet", "Crochet", "rue", "tape", 10, 12, rang=3, arc=1.4, temps=[
        _e("debout", 4, dx=-2, rot=-0.15), _e("crochet", 4, dx=2, rot=0.2, actif=True), _e("crochet", 4),
        _e("debout", 3)]),
    _t("genou", "Coup de genou", "rue", "genou", 10, 9, temps=[
        _e("debout", 3, dy=-1), _e("genou", 5, dx=2, z=2, actif=True), _e("debout", 4)]),
    _t("repousser", "Repousser", "rue", "prise", 0, 14, assomme=False, sans_sang=True, temps=[
        _e("saisie", 6, dx=3, actif=True), _e("debout", 4)]),
    _t("genoux_prise", "Genoux", "rue", "prise_frappe", 8, 14, temps=[
        _e("saisie", 2), _e("genou", 5, dx=1, z=2, actif=True), _e("saisie", 3)]),
    # --- Boxe
    _t("uppercut", "Uppercut", "boxe", "tape", 14, 12, rang=4, prix=300, renverse=True, temps=[
        _e("accroupi", 5, dy=1), _e("crochet", 4, z=3, actif=True), _e("crochet", 4, z=1), _e("debout", 4)]),
    # --- Karaté : plus loin, plus lent.
    _t("pied_circulaire", "Coup de pied circulaire", "karate", "tape", 14, 18, rang=5, prix=500,
       arc=2.4, renverse=True, temps=[
        _e("debout", 5, rot=-0.3), _e("pied_cote", 5, rot=0.6, actif=True), _e("pied_cote", 4, rot=1.2),
        _e("debout", 5)]),
    _t("pied_de_cote", "Coup de pied de côté", "karate", "tenue", 16, 20, prix=400, arc=0.8,
       renverse=True, projete=40, tours=0.0, temps=[
        _e("genou", 8, dx=-1), _e("pied_cote", 6, dx=3, actif=True), _e("pied_cote", 6, dx=2), _e("debout", 6)]),
    _t("pied_saute", "Coup de pied sauté", "karate", "course", 14, 22, prix=600, renverse=True, temps=[
        _e("accroupi", 4), _e("pied_face", 6, dx=6, z=6, actif=True), _e("pied_face", 4, dx=3, z=3),
        _e("accroupi", 5)]),
    _t("balayage", "Balayage", "karate", "roulade", 6, 16, prix=450, arc=6.3, renverse=True, temps=[
        _e("accroupi", 3, dy=2), _e("accroupi", 8, dy=2, actif=True), _e("debout", 5)]),
    # --- Judo : les projections. Pas de sang, et on assomme.
    _t("projection_hanche", "Projection de hanche", "judo", "prise_avant", 12, 14, prix=500,
       renverse=True, projete=28, sans_sang=True, temps=[
        _e("saisie", 4), _e("hanche", 10, dx=2, actif=True), _e("hanche", 6), _e("debout", 6)]),
    _t("grand_fauchage", "Grand fauchage", "judo", "prise_vers_soi", 12, 14, prix=500,
       renverse=True, projete=6, sans_sang=True, temps=[
        _e("saisie", 4), _e("pied_face", 8, dx=2, actif=True), _e("saisie", 6), _e("debout", 5)]),
    _t("sacrifice", "Sacrifice en cercle", "judo", "prise_cote", 14, 14, prix=700,
       renverse=True, projete=40, tours=1.0, sans_sang=True, temps=[
        _e("saisie", 4), _e("au_sol", 14, actif=True), _e("accroupi", 8), _e("debout", 4)]),
    # --- Jiu-jitsu
    _t("retournement_poignet", "Retournement du poignet", "jiujitsu", "contre", 12, 20, prix=800,
       renverse=True, projete=16, sans_sang=True, temps=[
        _e("saisie", 10, actif=True), _e("accroupi", 6), _e("debout", 4)]),
    _t("etranglement", "Étranglement", "jiujitsu", "dos", 999, 14, prix=900, silencieuse=True,
       sans_sang=True, tenir=90, temps=[
        _e("saisie", 90, actif=True), _e("debout", 6)]),
]


def par_slug(slug: str) -> Technique:
    return next(t for t in CATALOGUE if t["slug"] == slug)


def chaine() -> list[Technique]:
    """Les techniques de la chaîne de tapes, dans l'ordre des rangs."""
    return sorted((t for t in CATALOGUE if t["geste"] == "tape"), key=lambda t: t["rang"])

File: static/js/techniques.js
/* Bandini — les techniques d'arts martiaux : la chaine de tapes, les pieds,
   les projections (docs/jalons/les-techniques-d-arts-martiaux.md).

   Python decide (`app/techniques.py` : le catalogue, les lignes de temps), ce
   module JOUE : il choisit la technique d'apres le geste et le contexte, fait
   avancer sa ligne de temps, et applique le coup a l'etape `actif`.

   ⚠️ AUCUN `B.rng()` ICI. Un passant choisit sa variante a l'EMPREINTE (son id,
   son compteur de coups) : un de consomme ici decalerait tout le hasard de la
   ville (voir les juges « ce module ne deplace rien »). */

const Techniques = (function () {
  'use strict';

  //: La fenetre de rythme : une tape dans les 0,4 s apres la fin d'un coup
  //: passe au maillon suivant ; au-dela, on repart du direct du gauche.
  const FENETRE = 24;
  //: Colle a la cible (en px) : le genou prend la place du crochet. ⚠️ Au
  //: CONTACT, pas plus pres : deux corps ne s'approchent jamais sous 10 px
  //: (`Entites.demeler`) — a 9, le genou ne partait jamais.
  const COLLE = 11;
  //: Le maillon que le genou remplace quand on est colle.
  const RANG_DU_GENOU = 3;
  //: Au sortir d'une roulade, combien d'images FRAPPE donne encore le balayage.
  const SORTIE_ROULADE = 12;

  function def(slug) { return (B.defs.techniques || []).find(function (t) { return t.slug === slug; }) || null; }

  /** `e` connait-il `slug` ? La rue, tout le monde ; le reste, ce que la partie a appris. */
  function sait(e, slug) {
    const t = def(slug);
    if (!t) return false;
    if (t.gratuite) return true;
    if (Entites.estJoueur(e)) return !!(B.partie && B.partie.techniques && B.partie.techniques[slug]);
    return !!(e.techniques && e.techniques.indexOf(slug) >= 0);
  }

  function parGeste(geste, rang) {
    return (B.defs.techniques || []).find(function (t) { return t.geste === geste && (!rang || t.rang === rang); }) || null;
  }

  /** La personne la plus proche de `e` dans `rayon`, ou null. */
  function cibleProche(e, rayon) {
    let mieux = null, d2 = rayon * rayon;
    for (const c of Entites.autour(e.x, e.y, rayon, function (q) {
      return q !== e && q.vivant && (q.type === 'pieton' || q.type === 'joueur');
    })) {
      const d = dist2(e.x, e.y, c.x, c.y);
      if (d <= d2) { d2 = d; mieux = c; }
    }
    return mieux;
  }

  /** Pure : le slug que `geste` donne a `e`, ou null s'il ne sait pas le faire.
      Pour une tape : le maillon suivant de la chaine s'il est su et que la
      fenetre de rythme court, sinon le direct du gauche. */
  function choisir(e, geste) {
    if (geste === 'tape') {
      const suivant = e.chaineT > 0 ? (e.chaine || 0) + 1 : 1;
      const t = parGeste('tape', suivant);
      return t && sait(e, t.slug) ? t.slug : parGeste('tape', 1).slug;
    }
    const t = parGeste(geste);
    return t && sait(e, t.slug) ? t.slug : null;
  }

  /** Le geste de FRAPPE, d'apres le contexte : tenu, au sortir d'une roulade,
      en sprint, colle a la cible — sinon une tape. */
  function gesteDeFrappe(e, fort) {
    if (fort) return 'tenue';
    if (e.sortieRoulade > 0) return 'roulade';
    const ent = e.entree || Entree;
    const v = B.defs.recherche.vitesses;
    if (ent.bas('esquive') && Math.hypot(e.vx, e.vy) > v.joueur_course + 0.1) return 'course';
    return 'tape';
  }

  /** Lance une tape de la chaine ; collé a la cible, le crochet devient le genou
      (et la chaine continue derriere lui). */
  function lancerTape(e, slug) {
    const t = def(slug);
    if (t.rang === RANG_DU_GENOU && cibleProche(e, COLLE)) {
      if (!demarrer(e, 'genou', null)) return false;
      e.chaine = RANG_DU_GENOU; e.chaineT = 0;
      return true;
    }
    return demarrer(e, slug, null);
  }

  /** Remplace `Combat.frapper` a mains nues. Rend true si un coup part. */
  function frapper(e, fort) {
    // ⚠️ UNE TAPE PENDANT LE COUP N'EST PAS PERDUE : on la garde (une seule), et
    // elle part a la fin du coup en cours. Sans elle, on ratait une tape sur deux.
    if (e.etat === 'attaque' && e.technique) { e.reserve = true; return false; }
    if (e.etat === 'attaque' || e.roule > 0) return false;
    if (!Entites.estJoueur(e)) {
      // La rue seulement, a l'empreinte : gauche, droit, crochet — le genou colle.
      e.coups = (e.coups || 0) + 1;
      return lancerTape(e, parGeste('tape', ((e.id + e.coups) % 3) + 1).slug);
    }
    const geste = gesteDeFrappe(e, fort);
    const slug = choisir(e, geste);
    if (slug) return geste === 'tape' ? lancerTape(e, slug) : demarrer(e, slug, null);
    // Sans le cours : une tape. TENUE, c'est le coup fort d'avant, sur le maillon.
    if (!lancerTape(e, choisir(e, 'tape'))) return false;
    if (geste === 'tenue') e.fort = true;
    return true;
  }

  /** Ce que `Combat.arcDeMelee` sait lire : une arme de la forme d'`armes.py`. */
  function armeDe(t, e) {
    // Le poing americain garde son ecart d'`armes.py` (12 contre 8) : +4.
    const lest = e && e.arme === 'poing_americain' ? 4 : 0;
    return { slug: e && e.arme === 'poing_americain' ? 'poing_americain' : 'poings', type: 'melee',
             degats: t.degats + lest, portee: t.portee, arc: t.arc, renverse: t.renverse, saigne: 0,
             assomme: t.assomme, usures: 0, son: 'coup', sans_sang: t.sans_sang,
             silencieuse: t.silencieuse, technique: t.slug };
  }

  /** Lance `slug` sur `e` (et sur `cible`, pour une prise). */
  function demarrer(e, slug, cible) {
    const t = def(slug);
    if (!t) return false;
    // ⚠️ ON SE SOUVIENT DE CE QU'ON FAISAIT (voir `Combat.frapper`) : un passant
    // qui se battait avec l'autre gang y retourne apres son coup.
    if (e.etat !== 'attaque') e.avantLeCoup = e.etat;
    e.etat = 'attaque';
    e.technique = slug;
    e.techCible = cible || null;
    e.techEtape = 0;
    e.techT = t.temps[0].images;
    e.techPose = t.temps[0].pose;
    e.touches = [];
    e.fort = false;
    e.reserve = false;
    e.phase = t.temps[0].actif ? 'actif' : 'anticipation';
    e.arc = armeDe(t, e);
    if (t.geste === 'tape') { e.chaine = t.rang; e.chaineT = 0; }
    // Le retournement, l'etranglement : l'etape 0 est deja l'actif.
    if (t.temps[0].actif) entrerDansLActif(e, t);
    return true;
  }

  /** Les gestes qui ne frappent pas en arc : la prise porte sur SA cible. */
  function frappeEnArc(t) {
    return t.geste !== 'contre' && t.geste !== 'dos' && t.geste.indexOf('prise') !== 0;
  }

  /** Une image de la ligne de temps de `e`. */
  function maj(e) {
    const t = def(e.technique);
    if (!t) { finir(e); return; }
    if (t.temps[e.techEtape].actif && frappeEnArc(t)) Combat.arcDeMelee(e, e.arc);
    if (--e.techT > 0) return;
    e.techEtape++;
    if (e.techEtape >= t.temps.length) { finir(e); return; }
    const suivante = t.temps[e.techEtape];
    e.techT = suivante.images;
    e.techPose = suivante.pose;
    const iActif = t.temps.findIndex(function (q) { return q.actif; });
    e.phase = suivante.actif ? 'actif' : (e.techEtape > iActif ? 'repos' : 'anticipation');
    if (suivante.actif) entrerDansLActif(e, t);
  }

  function entrerDansLActif(e, t) {
    const arme = e.arc;
    Son.depuis(e, function () { Son.SFX.arme(arme); });
    // La prise branche ici la projection de `e.techCible`. ⚠️ `api`, pas
    // `Techniques` : appele a l'execution, apres la fin de l'IIFE.
    if (api.surActif) api.surActif(e, t);
  }

  function finir(e) {
    const t = def(e.technique);
    const reserve = e.reserve;
    e.technique = null; e.techPose = null; e.techCible = null; e.phase = null; e.reserve = false;
    const enChaine = t && (t.geste === 'tape' || t.geste === 'genou');
    if (enChaine) e.chaineT = FENETRE;
    if (Entites.estJoueur(e)) {
      e.etat = 'flane';
      if (reserve && enChaine) frapper(e, false);
      return;
    }
    e.etat = e.avantLeCoup || 'attaque_joueur';
    e.avantLeCoup = null;
  }

  /** Les compteurs qui courent hors du coup : la fenetre de la chaine, la sortie de roulade. */
  function majCompteurs(e) {
    if (e.chaineT > 0 && e.etat !== 'attaque') e.chaineT--;
    if (e.sortieRoulade > 0 && !(e.roule > 0)) e.sortieRoulade--;
  }

  const api = { FENETRE, COLLE, SORTIE_ROULADE, def, sait, choisir, frapper, demarrer, maj, majCompteurs,
                cibleProche, surActif: null };
  return api;
})();
